controllers/helpers.go

```diff
@@ -777,5 +787,182 @@ func getPrismCentralClientForCluster(ctx context.Context, cluster *infrav1.Nutan
+	parts := strings.Split(version, ".")
+	if len(parts) < 1 {
+		return false, errors.New("invalid version format")
+	}
+
+	// Convert the year part to an integer
+	year, err := strconv.Atoi(parts[0])
+	if err != nil {
+		return false, errors.New("invalid version: failed to parse year from PC version")
+	}
+
+	if year >= 2024 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// getPrismCentralVersion returns the version of the Prism Central instance
+func getPrismCentralVersion(ctx context.Context, v3Client *prismclientv3.Client) (string, error) {
+	pcInfo, err := v3Client.V3.GetPrismCentral(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	if pcInfo.Resources == nil || pcInfo.Resources.Version == nil {
+		return "", fmt.Errorf("failed to get Prism Central version")
+	}
+
+	return *pcInfo.Resources.Version, nil
+}
+
+func detachVolumeGroupsFromVM(ctx context.Context, v4Client *prismclientv4.Client, vmUUID string) error {
+	log := ctrl.LoggerFrom(ctx)
+	vmResp, err := v4Client.VmApiInstance.GetVmById(&vmUUID)
+	if err != nil {
+		return fmt.Errorf("failed to get virtual machine: %w", err)
+	}
+
+	data := vmResp.GetData()
+	vm, ok := data.(vmconfig.Vm)
+	if !ok {
+		return fmt.Errorf("failed to cast response to VM")
+	}
+
+	volumeGroupsToDetach := make([]string, 0)
+	for _, disk := range vm.Disks {
+		backingInfo, ok := disk.GetBackingInfo().(vmconfig.ADSFVolumeGroupReference)
+		if !ok {
+			continue
+		}
+
+		volumeGroupsToDetach = append(volumeGroupsToDetach, *backingInfo.VolumeGroupExtId)
+	}
+
+	log.Info(fmt.Sprintf("detaching %d volume groups from virtual machine %s", len(volumeGroupsToDetach), vmUUID))
+
+	// Detach the volume groups from the virtual machine
+	for _, volumeGroup := range volumeGroupsToDetach {
+		log.Info(fmt.Sprintf("detaching volume group %s from virtual machine %s", volumeGroup, vmUUID))
+		body := &volumesconfig.VmAttachment{
+			ExtId: &vmUUID,
+		}
+		resp, err := v4Client.VolumeGroupsApiInstance.DetachVm(&volumeGroup, body)
+		if err != nil {
+			return fmt.Errorf("failed to detach volume group %s from virtual machine %s: %w", volumeGroup, vmUUID, err)
+		}
+
+		data := resp.GetData()
+		task, ok := data.(prismconfig.TaskReference)
+		if !ok {
+			return fmt.Errorf("failed to cast response to TaskReference")
+		}
+
+		// Wait for the task to complete
+		if _, err := waitForTaskCompletionV4(ctx, v4Client, *task.ExtId); err != nil {
+			return fmt.Errorf("failed to wait for task %s to complete: %w", *task.ExtId, err)
+		}
+	}
+
+	return nil
+}
+
+// waitForTaskCompletionV4 waits for a task to complete and returns the completion details
+// of the task. The function will poll the task status every 100ms until the task is
+// completed or the context is cancelled.
+func waitForTaskCompletionV4(ctx context.Context, v4Client *prismclientv4.Client, taskID string) ([]prismcommonconfig.KVPair, error) {
+	var data []prismcommonconfig.KVPair
+
+	if err := wait.PollUntilContextCancel(
```

Comment:
what happens when the task is say hung. instead of getting stuck here, would storing the taskId in the status field and returning with retry from reconcile and letting k8s call back would be better? we dont have to remove the finalizer till the task is complete but we avoid loooping ourselves by holding off main control loop.

Reply:
If the task is hung, it will wait on the task until context gets cancelled. This isn't much different than what we do with v3 task wait. The idea is that we get synchronous feedback when a task completes before executing the next bit which in the context of this PR is detaching the VG and deleting the VM.